service/lib/y2storage/proposal/agama_device_planner.rb

```diff
@@ -0,0 +1,124 @@
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "y2storage/planned"
+
+module Y2Storage
+  module Proposal
+    # Base class used by Agama planners.
+    class AgamaDevicePlanner
+      # @!attribute [r] devicegraph
+      #   @return [Devicegraph]
+      attr_reader :devicegraph
+
+      # @!attribute [r] config
+      #   @return [Agama::Config]
+      attr_reader :config
+
+      # @!attribute [r] issues_list
+      attr_reader :issues_list
+
+      # @param devicegraph [Devicegraph] Devicegraph to be used as starting point.
+      # @param issues_list [AutoinstIssues::List] List of issues to register them.
+      def initialize(devicegraph, config, issues_list)
+        @devicegraph = devicegraph
+        @config = config
+        @issues_list = issues_list
+      end
+
+      # Planned devices according to the given settings.
+      #
+      # @return [Array] Array of planned devices.
+      def planned_devices(_setting)
+        raise NotImplementedError
+      end
+
+      private
+
+      # @param planned [Planned::Disk, Planned::Partition]
+      # @param settings [#format, #mount]
+      def configure_device(planned, settings)
+        # TODO configure_encrypt
+        configure_format(planned, settings.format) if settings.format
+        configure_mount(planned, settings.mount) if settings.mount
+      end
+
+      # @param planned [Planned::Disk, Planned::Partition]
+      # @param settings [Agama::Storage::Settings::Format]
+      def configure_format(planned, settings)
+        planned.label = settings.label
+        planned.mkfs_options = settings.mkfs_options
+        configure_filesystem(planned, settings.filesystem) if settings.filesystem
+      end
+
+      # @param planned [Planned::Disk, Planned::Partition]
+      # @param settings [Agama::Storage::Settings::Filesystem]
+      def configure_filesystem(planned, settings)
+        planned.filesystem_type = settings.type
+        configure_btrfs(planned, settings.btrfs) if settings.btrfs
+      end
+
+      # @param planned [Planned::Disk, Planned::Partition]
+      # @param settings [Agama::Storage::Settings::Btrfs]
+      def configure_btrfs(planned, settings)
+        planned.snapshots = settings.snapshots?
+        planned.default_subvolume = settings.default_subvolume
+        planned.subvolumes = settings.subvolumes
+      end
+
+      # @param planned [Planned::Disk, Planned::Partition]
+      # @param settings [Agama::Storage::Settings::Mount]
+      def configure_mount(planned, settings)
+        planned.mount_point = settings.path
+        planned.mount_by = settings.mount_by
+        planned.fstab_options = settings.options
+        # FIXME: Is this needed? Or #options is enough?
+        # planned.read_only = settings.read_only?
```

Comment:
Checking how this `read_only` is used so far. For both the Guided and the AutoYaST proposals, this is used only for Btrfs planned filesystems, honoring `VolumeSpecification#btrfs_read_only`.

```ruby
class VolumeSpecification
  attr_accessor :btrfs_read_only
end

module Proposal
  class DevicesPlanner
    def adjust_btrfs(planned_device, volume)
      return unless planned_device.btrfs?
      
      #[...]
      planned_device.read_only = volume.btrfs_read_only?
    end 
  end 
end 

module Proposal
  class AutoinstDrivePlanner
    def add_filesystem_attrs(device, partition_section)
      #[...]
      device.read_only = read_only?(partition_section.mount)
    end

    def read_only?(mount_point)
      return false unless mount_point

      spec = VolumeSpecification.for(mount_point)
      !!spec && spec.btrfs_read_only?
    end
  end
end
```

When committing it to the system, the `read_only` flag is indeed translated to the `fstab_options`.

```ruby
module Planned
  module CanBeMounted
    attr_accessor :read_only
  end
end

module Planned
  module CanBeFormatted
    def fstab_options_for(mount_point)
      #[...]
      options.push("ro") if read_only && !options.include?("rw")
      #[...]
    end
  end
end
```

Reply:
We moved the topic a bit forward to https://trello.com/c/4Cm7lh1j/492-storage-profile-filesystem-related-properties-especially-btrfs-ones